Check/uftrace print

honggyukim opened the pull request:
uftrace is able to capture function arguments and return values and that is the core feature.

But sometimes, it's also necessary to see some internal variables or even user may want to put a marker to distinguish the point from a long sequence of function calls.

This PR is an implementation of #428 so please see the issue for more details.

The usage is as follows:
1. Include "uftrace_print.h" in the target source code.
2. Modify the source code by inserting uftrace_print() function with a
    single argument that you want to record with its function flow.
3. Run "uftrace record" for the target program.
4. Run "uftrace replay" to observe the values.

Here is an example. Let's say `tests/s-arg.c` is modified as follows:
```diff
$ diff -u tests/s-arg.c tests/s-arg-uftrace_print.c
--- tests/s-arg.c     2018-12-27 22:23:42.142927670 +0900
+++ tests/s-arg-uftrace_print.c       2018-12-27 21:12:46.590220776 +0900
@@ -6,6 +6,7 @@
 #include <stdarg.h>
 #include <stdlib.h>
 #include <string.h>
+#include <uftrace_print.h>

 const char *strs[] = { "a", "b", "c" };
 const unsigned nr_strs = sizeof(strs)/sizeof(strs[0]);
@@ -61,6 +62,13 @@

 int check(struct big val, struct big *ref)
 {
+       uftrace_print(val.c);
+       uftrace_print(ref->d);
+       uftrace_print(val.s);
+       uftrace_print(ref->ld);
+       uftrace_print(val.i);
+       uftrace_print(ref->ull);
+       uftrace_print(val.l);
        if (val.c != ref->c)
                return -1;
        if (val.d != ref->d)
@@ -81,6 +89,7 @@
 int pass(int n)
 {
        struct big b = { 'b', 3.14, -1, 2.71828, n, 987654321ULL, 12345L };
+       uftrace_print(&b);
        return check(b, &b);
 }
```
Then, if it's traced with uftrace, it records the parameter values of uftrace_print as follows:
```
$ gcc -pg -g tests/s-arg-uftrace_print.c
$ uftrace -a -F pass a.out
# DURATION     TID     FUNCTION
            [129490] | pass(3) {
 319.995 us [129490] |   uftrace_print_pv("&b", 0x7ffeddcd84c0);
            [129490] |   check(0x7ffeddcd84c0, 0x15ca470) {
   0.470 us [129490] |     uftrace_print_c("val.c", 'b');
   0.453 us [129490] |     uftrace_print_d("ref->d", 3.140000);
   0.403 us [129490] |     uftrace_print_s("val.s", -1);
   0.333 us [129490] |     uftrace_print_ld("ref->ld", 0x15ca4c8);
   0.270 us [129490] |     uftrace_print_i("val.i", 3);
   0.326 us [129490] |     uftrace_print_ull("ref->ull", 0x3ade68b1);
   0.274 us [129490] |     uftrace_print_l("val.l", 12345);
   7.260 us [129490] |   } = 0; /* check */
 331.649 us [129490] | } = 0; /* pass */
```
It can handles both `C` and `C++` programs.

honggyukim commented:
After testing it again, `uftrace_print_*` functions have to be implemented inside `libmcount*.so` to avoid duplication. `uftrace_print.h` is supported to have their declarations only.

namhyung commented on uftrace_print.h:
```diff
@@ -0,0 +1,101 @@
+void uftrace_print_cxx(const char *X, const char* s)
+{
+    volatile const char c = *s;
+}
```
Maybe you could define it as a static function to prevent multiple definition error.

honggyukim replied:
I just changed it to "static" function. Thanks.

namhyung commented:
It seems gcc 4.9 added the `_Generic` support and it's enabled by default from gcc 5.x.  And later compilers could use older standard for some reason.  Isn't it better checking the `__STDC_VERSION__` and show warning?

honggyukim commented on tests/s-arg-uftrace_print.c:
```diff
@@ -0,0 +1,110 @@
+#include <stdarg.h>
+#include <stdlib.h>
+#include <string.h>
+#include <uftrace_print.h>
```
I added include base directory to make it work without `install`.

honggyukim commented on uftrace_print.h:
```diff
@@ -0,0 +1,101 @@
+#if defined(__clang__)
+#define UFT_DISABLE_OPT __attribute__((optnone))
+#elif defined(__GNUC__)
+#define UFT_DISABLE_OPT __attribute__((optimize(0)))
```
Hmm.. I made this change long time ago, so I don't remember clearly. But as far as I remember, there was a reason not to use `noinline` here. I will tell you when I find the reason.

honggyukim commented:
I just updated as follows:
- Add `__STDC_VERSION__` guard to protect `_Generic` and make a warning when compiler doesn't support `c11` yet.
- Fix `uftrace_print_*` functions to simply `uftrace_print` to hide unnecessary details in `utils/symbol.c`.
- Change test file names removing `s-arg-uftrace_print.*` to `s-uftrace_print.*`.
- Add include base directory to test cases.

I still have more things to consider so this is an interim update. Thanks for your feedback.

namhyung commented on tests/runtest.py:
```diff
@@ -32,6 +32,8 @@ class TestBase:
 
     def __init__(self, name, result, lang='C', cflags='', ldflags='', sort='task'):
+        # assume that this script is executed in uftrace/tests directory.
+        self.srcdir = os.path.dirname(os.getcwd())
```
The assumption is correct otherwise it cannot find any test case.

honggyukim replied:
Would it be better to remove the comment?

namhyung replied:
I don't care :)

honggyukim commented:
> I don't want to leave the header file in the root directory. Maybe 'misc'?

Okay. I will move it into `misc` directory.

> Otherwise it looks ok but it needs documentation.

Sure. I will write a document when I have time.

> And github seems to mix the order of commits in PR and it makes me annoying.

I also think so. :(

honggyukim commented:
In addition to providing `uftrace_print` functions, it'd be better to add some additional feature to this common header file.  The possible candidates are as follows:
- Provide functions for triggering `trace-on`/`trace-off` and `finish` without command line input. For example, `uftrace_trace_on`, `uftrace_trace_off` and `uftrace_trace_finish`.
- Provide a weak declaration to `mcount` function so that user can add `mcount` call directly to their source code. Users can add `mcount` call at the entry of function that they want to trace. It can be used for non `-pg` compiled binary.
- Provide C++ class, `uftrace_block(char *name)`, that can be used as a scoped range. For example, chrome uses this method to add SDT. They usually get timestamp both at constructors and destructors to calculate duration for the block.

I would like to get your feedback. Thanks!

honggyukim commented:
Pushed again for interim update.
- Rename and move `uftrace_print.h` to `include/uftrace.h`
- Add include guard and license statement to `include/uftrace.h`
- Add `__attribute__((unused))` to prevent unused variable warning

honggyukim commented:
Hmm.. I also have to handle the below case.
```
$ uftrace -a -N uftrace_print a.out
```
It doesn't hide `uftrace_print` functions.

honggyukim commented:
@namhyung Can you please review 7d810d1b first?

namhyung commented:
Could you please check [check/demangle-fix](https://github.com/namhyung/uftrace/compare/check/demangle-fix)?

honggyukim commented:
After removing 80db5c0, then rebased based on [check/demangle-fix](https://github.com/namhyung/uftrace/compare/check/demangle-fix), but it doesn't solve the problem.
```
$ uftrace -a t-uftrace_print
# DURATION     TID     FUNCTION
   1.534 us [ 58374] | __monstartup();
   1.140 us [ 58374] | __cxa_atexit();
            [ 58374] | main() {
            [ 58374] |   pass(3) {
   0.230 us [ 58374] |     uftrace_print();
            [ 58374] |     check(0x7fffc1d278d0, 0x217b968) {
   0.177 us [ 58374] |       uftrace_print();
   0.147 us [ 58374] |       uftrace_print();
   0.143 us [ 58374] |       uftrace_print();
   0.133 us [ 58374] |       uftrace_print();
   0.143 us [ 58374] |       uftrace_print();
   0.130 us [ 58374] |       uftrace_print();
   0.137 us [ 58374] |       uftrace_print();
   0.137 us [ 58374] |       uftrace_print();
   0.133 us [ 58374] |       uftrace_print();
   0.133 us [ 58374] |       uftrace_print();
   6.057 us [ 58374] |     } = 0; /* check */
  10.427 us [ 58374] |   } = 0; /* pass */
  11.620 us [ 58374] | } = 0; /* main */
```
The parameter values can be shown only when demangling is off.
```
$ uftrace -a --demangle=no t-uftrace_print
# DURATION     TID     FUNCTION
   1.557 us [ 58396] | __monstartup();
   1.077 us [ 58396] | __cxa_atexit();
            [ 58396] | main() {
            [ 58396] |   pass(3) {
 319.522 us [ 58396] |     _ZL13uftrace_printPKcPKvz("&b", 0x7ffc03d2b890);
            [ 58396] |     check(0x7ffc03d2b890, 0xd8cad8) {
   0.510 us [ 58396] |       _ZL13uftrace_printPKcc("val.c", 'b');
   0.547 us [ 58396] |       _ZL13uftrace_printPKcd("ref->d", 3.140000);
   0.357 us [ 58396] |       _ZL13uftrace_printPKcs("val.s", -1);
   0.330 us [ 58396] |       _ZL13uftrace_printPKce("ref->ld", 0xd8cb30);
   0.293 us [ 58396] |       _ZL13uftrace_printPKci("val.i", 3);
   0.323 us [ 58396] |       _ZL13uftrace_printPKcy("ref->ull", 0x3ade68b1);
   0.290 us [ 58396] |       _ZL13uftrace_printPKcl("val.l", 12345);
   0.467 us [ 58396] |       _ZL13uftrace_printPKcPc(""string test!"", "string test!");
   0.376 us [ 58396] |       _ZL13uftrace_printPKcPc("val.str", "hello");
   0.363 us [ 58396] |       _ZL13uftrace_printPKcPc("ref->str", "hello");
   9.097 us [ 58396] |     } = 0; /* check */
 332.603 us [ 58396] |   } = 0; /* pass */
 333.703 us [ 58396] | } = 0; /* main */
```

honggyukim commented:
It shows the arguments of `uftrace_print` properly now. Thanks!

honggyukim commented:
I will rebase and update again if you merge the branch `check/demangle-fix`.

honggyukim commented:
I just rebased and added one more commit that has the following features:
- `mcount()` declaration for explicit function tracing without -pg
- `uftrace_trace_on()` to enable "trace_on" trigger
- `uftrace_trace_off()` to enable "trace_off" trigger
- `uftrace_finish()` to enable "finish" trigger

Explicit `mcount()` works fine as is, but it requires to modify uftrace internal to make other trigger functions work. It's just a preparation work.

honggyukim commented:
I just added default triggers for `uftrace_trace_on`, `uftrace_trace_off`, and `uftrace_finish`.
For testing, I modified the a test example as follows:
```diff
$ git diff
diff --git a/tests/s-uftrace_print.c b/tests/s-uftrace_print.c
index 89cd381..ee5be63 100644
--- a/tests/s-uftrace_print.c
+++ b/tests/s-uftrace_print.c
@@ -23,11 +23,14 @@ int check(struct big val, struct big *ref)
        uftrace_print(ref->d);
        uftrace_print(val.s);
        uftrace_print(ref->ld);
+       uftrace_trace_off();
        uftrace_print(val.i);
        uftrace_print(ref->ull);
        uftrace_print(val.l);
+       uftrace_trace_on();
        uftrace_print("string test!");
        uftrace_print(val.str);
+       uftrace_finish();
        uftrace_print(ref->str);
        return 0;
 }
@@ -35,6 +38,7 @@ int check(struct big val, struct big *ref)
 int pass(int n)
 {
        struct big b = { 'b', 3.14, -1, 2.71828, "hello", n, 987654321ULL, 12345L };
+       uftrace_trace_on();
        uftrace_print(&b);
        return check(b, &b);
 }
```
It's compiled again, then the trace output looks as follows:
```
$ gcc -pg -g s-uftrace_print.c

$ uftrace --disable -a a.out
# DURATION     TID     FUNCTION
   0.290 us [ 92733] |     uftrace_trace_on();
 322.187 us [ 92733] |     uftrace_print("&b", 0x7ffc1971daf0);
            [ 92733] |     check(0x7ffc1971daf0, 0x25d9d58) {
   0.514 us [ 92733] |       uftrace_print("val.c", 'b');
   0.566 us [ 92733] |       uftrace_print("ref->d", 3.140000);
   0.357 us [ 92733] |       uftrace_print("val.s", -1);
   0.337 us [ 92733] |       uftrace_print("ref->ld", 0x25d9db0);
   0.180 us [ 92733] |       uftrace_trace_on();
   0.460 us [ 92733] |       uftrace_print(""string test!"", "string test!");
   0.324 us [ 92733] |       uftrace_print("val.str", "hello");
            [ 92733] |       uftrace_finish() {
            [ 92733] |         /* linux:task-exit */

uftrace stopped tracing with remaining functions
================================================
task: 92733
[1] uftrace_finish
[0] check

```

honggyukim commented:
There's tiny depth difference between the output of `live` and `replay` as follows:
```
$ uftrace replay
# DURATION     TID     FUNCTION
   0.297 us [ 92794] | uftrace_trace_on();
 318.656 us [ 92794] | uftrace_print("&b", 0x7ffe4e79cad0);
            [ 92794] | check(0x7ffe4e79cad0, 0x200ad28) {
   0.484 us [ 92794] |   uftrace_print("val.c", 'b');
   0.646 us [ 92794] |   uftrace_print("ref->d", 3.140000);
   0.383 us [ 92794] |   uftrace_print("val.s", -1);
   0.320 us [ 92794] |   uftrace_print("ref->ld", 0x200ad80);
   0.156 us [ 92794] |   uftrace_trace_on();
   0.453 us [ 92794] |   uftrace_print(""string test!"", "string test!");
   0.366 us [ 92794] |   uftrace_print("val.str", "hello");
            [ 92794] |   uftrace_finish() {
            [ 92794] |     /* linux:task-exit */

uftrace stopped tracing with remaining functions
================================================
task: 92794
[1] uftrace_finish
[0] check
```
